web/ui/react-app/src/utils/urlParams.ts

```diff
@@ -1,127 +1,85 @@
+  const decodedValue = decodeURIComponent(val.replace(/\+/g, ' '));
   switch (opt) {
     case 'expr':
-      opts.expr = val;
-      break;
+      return { expr: decodedValue };
 
     case 'tab':
-      if (val === '0') {
-        opts.type = PanelType.Graph;
-      } else {
-        opts.type = PanelType.Table;
-      }
-      break;
+      return { type: decodedValue === '0' ? PanelType.Graph : PanelType.Table };
 
     case 'stacked':
-      opts.stacked = val === '1';
-      break;
+      return { stacked: decodedValue === '1' };
 
     case 'range_input':
-      const range = parseRange(val);
-      if (range !== null) {
-        opts.range = range;
-      }
-      break;
+      const range = parseRange(decodedValue);
+      return isPresent(range) ? { range } : {};
 
     case 'end_input':
-      opts.endTime = parseTime(val);
-      break;
+    case 'moment_input':
+      return { endTime: parseTime(decodedValue) };
 
     case 'step_input':
-      const res = parseInt(val);
-      if (res > 0) {
-        opts.resolution = res;
-      }
-      break;
-
-    case 'moment_input':
-      opts.endTime = parseTime(val);
-      break;
+      const resolution = parseInt(decodedValue);
+      return resolution > 0 ? { resolution } : {};
   }
-}
-
-export function encodePanelOptionsToQueryString(panels: { key: string; options: PanelOptions }[]): string {
-  const queryParams: string[] = [];
-
-  panels.forEach(p => {
-    const prefix = 'g' + p.key + '.';
-    const o = p.options;
-    const panelParams: { [key: string]: string | undefined } = {
-      expr: o.expr,
-      tab: o.type === PanelType.Graph ? '0' : '1',
-      stacked: o.stacked ? '1' : '0',
-      range_input: formatRange(o.range),
-      end_input: o.endTime !== null ? formatTime(o.endTime) : undefined,
-      moment_input: o.endTime !== null ? formatTime(o.endTime) : undefined,
-      step_input: o.resolution !== null ? o.resolution.toString() : undefined,
-    };
-
-    for (const o in panelParams) {
-      const pp = panelParams[o];
-      if (pp !== undefined) {
-        queryParams.push(prefix + o + '=' + encodeURIComponent(pp));
-      }
-    }
-  });
-
-  return '?' + queryParams.join('&');
-}
+  return {};
+};
+
+export const formatParam = (key: string) => (paramName: string, value: number | string | boolean) => {
+  return `g${key}.${paramName}=${encodeURIComponent(value)}`;
+};
+
+export const toQueryString = ({ key, options }: PanelMeta) => {
+  const formatWithKey = formatParam(key);
+  const { expr, type, stacked, range, endTime, resolution } = options;
+  const time = isPresent(endTime) ? formatTime(endTime) : false;
+  const urlParams = [
+    formatWithKey('expr', expr),
+    formatWithKey('tab', type === PanelType.Graph ? 0 : 1),
+    formatWithKey('stacked', stacked ? 1 : 0),
+    formatWithKey('range_input', formatRange(range)),
+    time ? `${formatWithKey('end_input', time)}&${formatWithKey('moment_input', time)}` : '',
```

Comment:
Hmm I know it's been this way before (in the new UI, not in classic UI), but I wonder if we actually want this behavior that the table and graph end times are synced. But ok to keep it like this for now.